flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -446,41 +446,39 @@ struct CharacterOpsBuilderImpl {
     // Note: Runtime for LEN_TRIM should also be available at some
     // point. For now use an inlined implementation.
     auto indexType = builder.getIndexType();
+    auto i1Type = mlir::IntegerType::get(1, builder.getContext());
     mlir::Value len = builder.createHere<fir::ConvertOp>(indexType, str.len);
     auto data = materializeValue(str);
     auto one = builder.createIntegerConstant(indexType, 1);
-    auto i1Type = mlir::IntegerType::get(1, builder.getContext());
+    auto minusOne = builder.createIntegerConstant(indexType, -1);
+    auto zero = builder.createIntegerConstant(indexType, 0);
     auto trueVal = builder.createIntegerConstant(i1Type, 1);
     auto blank = createBlankConstant(str.getCharacterType());
     // FIXME: Conversion wrong in general, it would be cool to compare fir.char
     // without cast. Needs the kindmap.
-    auto i8Type = mlir::IntegerType::get(8, builder.getContext());
-    blank = builder.createHere<fir::ConvertOp>(i8Type, blank);
+    auto intType = mlir::IntegerType::get(8, builder.getContext());
+    blank = builder.createHere<fir::ConvertOp>(intType, blank);
     // initVal must not be random to account for the zero length case where the
     // no iteration of the loop is ever executed.
     mlir::Value initVal = builder.createHere<mlir::SubIOp>(len, one);
 
     auto iterWhile =
-        builder.createHere<fir::IterWhileOp>(one, len, one, trueVal, initVal);
+        builder.createHere<fir::IterWhileOp>(initVal, zero, minusOne, trueVal, initVal);
     auto insPt = builder.saveInsertionPoint();
     builder.setInsertionPointToStart(iterWhile.getBody());
 
     auto index = iterWhile.getInductionVar();
-    auto charPosition = builder.createHere<mlir::SubIOp>(len, index);
-    auto c = createLoadCharAt(data, charPosition);
-    // FIXME: needs more generic conversion.
-    c = builder.createHere<fir::ConvertOp>(i8Type, c);
+    auto c = createLoadCharAt(data, index);
+    c = builder.createHere<fir::ConvertOp>(intType, c);
     auto isBlank =
         builder.createHere<mlir::CmpIOp>(mlir::CmpIPredicate::eq, blank, c);
-    llvm::SmallVector<mlir::Value, 2> results = {isBlank, charPosition};
+    llvm::SmallVector<mlir::Value, 2> results = {isBlank, index};
```

Comment:
You don't need to explicitly pass the induction variable back. It is returned automatically. (If not, it is a bug.)

Reply:
Thanks, that works.